plugin/pkg/scheduler/core/scheduling_queue.go

```diff
@@ -15,47 +15,694 @@ limitations under the License.
+func (p *PriorityQueue) AssignedPodUpdated(pod *v1.Pod) {
+	p.movePodsToActiveQueue(p.getPodsWithMatchingAffinityTerm(pod))
+}
+
+// MoveAllToActiveQueue moves all pods from unschedulableQ to activeQ. This
+// function adds all pods and then signals the condition variable to ensure that
+// if Pop() is waiting for an item, it receives it after all the pods are in the
+// queue and the head is the highest priority pod.
+// TODO(bsalamat): We should add a back-off mechanism here so that a high priority
+// pod which is unschedulable does not go to the head of the queue frequently. For
+// example in a cluster where a lot of pods being deleted, such a high priority
+// pod can deprive other pods from getting scheduled.
+func (p *PriorityQueue) MoveAllToActiveQueue() {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+	var unschedulablePods []interface{}
+	for _, pod := range p.unschedulableQ.pods {
+		unschedulablePods = append(unschedulablePods, pod)
+	}
+	p.activeQ.BulkAdd(unschedulablePods)
+	p.unschedulableQ.Clear()
+	p.receivedMoveRequest = true
+	p.cond.Broadcast()
+}
+
+func (p *PriorityQueue) movePodsToActiveQueue(pods []*v1.Pod) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+	for _, pod := range pods {
+		p.activeQ.Add(pod)
+		p.unschedulableQ.Delete(pod)
+	}
+	p.receivedMoveRequest = true
+	p.cond.Broadcast()
+}
+
+// getPodsWithMatchingAffinityTerm returns unschedulable pods which have
+// any affinity term that matches "pod".
+func (p *PriorityQueue) getPodsWithMatchingAffinityTerm(pod *v1.Pod) []*v1.Pod {
+	p.lock.RLock()
+	defer p.lock.RUnlock()
+	podsToMove := []*v1.Pod{}
+	for _, up := range p.unschedulableQ.pods {
+		affinity := up.Spec.Affinity
+		if affinity != nil && affinity.PodAffinity != nil {
+			terms := predicates.GetPodAffinityTerms(affinity.PodAffinity)
+			for _, term := range terms {
+				namespaces := priorityutil.GetNamespacesFromPodAffinityTerm(up, &term)
+				selector, err := metav1.LabelSelectorAsSelector(term.LabelSelector)
+				if err != nil {
+					glog.Errorf("Error getting label selectors for pod: %v.", up.Name)
+				}
+				if priorityutil.PodMatchesTermsNamespaceAndSelector(pod, namespaces, selector) {
+					podsToMove = append(podsToMove, up)
+				}
+			}
+		}
+	}
+	return podsToMove
+}
+
+// WaitingPodsOnNode returns pods that are nominated to run on the given node,
+// but they are waiting for other pods to be removed from the node before they
+// can be actually scheduled.
+func (p *PriorityQueue) WaitingPodsOnNode(nodeName string) []*v1.Pod {
+	return p.unschedulableQ.GetPodsForNode(nodeName)
+}
+
+// UnschedulablePodsMap holds pods that cannot be scheduled. This data structure
+// is used to implement unschedulableQ.
+type UnschedulablePodsMap struct {
+	// pods is a map key by a pod's full-name and the value is a pointer to the pod.
+	pods map[string]*v1.Pod
+	// nominatedPods is a map keyed by a node name and the value is a list of
+	// pods' full-names which are nominated to run on the node.
```

Comment:
When a pod is nominated, it is marked unschedulable. This is scheduler's default behavior to mark such pods as unschedulable since they cannot be scheduled right away. These pods are added to unschedulableQ and may be moved to the activeQ from time to time when we receive a move request. Scheduler tries to schedule them as soon as possible and does not wait to put them only on the same node as their nominated node.
In my second PR, there is a logic in checking whether a node can fit a pod that picks up the nominated pods from the queue and makes sure that the being tried pods fits in presence of these nominated pods.